src/mlpack/core/tree/rectangle_tree/minimal_coverage_sweep_impl.hpp

```diff
@@ -0,0 +1,260 @@
+/**
+ * @file minimal_coverage_sweep_impl.hpp
+ * @author Mikhail Lozhnikov
+ *
+ * Implementation of the MinimalCoverageSweep class, a class that finds a
+ * partition of a node along an axis.
+ */
+#ifndef MLPACK_CORE_TREE_RECTANGLE_TREE_MINIMAL_COVERAGE_SWEEP_IMPL_HPP
+#define MLPACK_CORE_TREE_RECTANGLE_TREE_MINIMAL_COVERAGE_SWEEP_IMPL_HPP
+
+#include "minimal_coverage_sweep.hpp"
+
+namespace mlpack {
+namespace tree {
+
+template<typename SplitPolicy>
+template<typename TreeType>
+typename TreeType::ElemType MinimalCoverageSweep<SplitPolicy>::
+SweepNonLeafNode(const size_t axis,
+                 const TreeType* node,
+                 typename TreeType::ElemType& axisCut)
+{
+  typedef typename TreeType::ElemType ElemType;
+
+  std::vector<SortStruct<ElemType>> sorted(node->NumChildren());
+
+  for (size_t i = 0; i < node->NumChildren(); i++)
+  {
+    sorted[i].d = SplitPolicy::Bound(node->Child(i))[axis].Hi();
+    sorted[i].n = i;
+  }
+  // Sort high bounds of children.
+  std::sort(sorted.begin(), sorted.end(), StructComp<ElemType>);
+
+  size_t splitPointer = fillFactor * node->NumChildren();
+
+  axisCut = sorted[splitPointer - 1].d;
+
+  // Check if the partition is suitable.
+  if (!CheckNonLeafSweep(node, axis, axisCut))
```

Comment:
Why are you doing this check first before iterating over different values of `splitPointer`?  I guess I don't understand the reasoning.  It seems like this entire block of code is just to check that there does exist a valid split on this axis, so I guess the first `if` is just a shortcut to avoid checking every possible cut.  If by default you want to go with a midpoint split, you can hardcode the `0.5` and remove `fillFactor`, I agree with that.  But I guess if the default split does not work, you are trying all other possible splits until you find one that works?